query/physicalplan/sampler.go

```diff
@@ -23,11 +34,41 @@ type ReservoirSampler struct {
 	i float64 // i is the current row number being sampled
 }
 
+type sample struct {
+	// Record is a pointer to the record that is being sampled
+	arrow.Record
+	// i is the index of the row in the record that is being sampled. If i is -1, the entire record is being sampled.
+	i int64
+	// ref is the number of references to the record. When ref is 0, the record can be released.
+	ref *int64
```

Comment:
It's a pointer because multiple samples reference the same underlying record. We don't want to pass around pointers to samples because in testing those escaped to heap and caused a performance hit.